### PR TITLE
passwordstore: Support subkey creation and update

### PR DESCRIPTION
**Pull request by mluzarreta**
##### SUMMARY

passwordstore lookup - Add the ability to create and update subkeys in password files.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

passwordstore lookup

##### ADDITIONAL INFORMATION

Here some command lines to  illustrate the feature:

```
# Create a new 'foo' subkey, in an new password store file (in this case, main password is empty)
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, subkey='foo') }}" localhost
localhost | SUCCESS => {
    "msg": "u6V31gj3d0i3lLbO"
}
# Replace the 'foo' subkey and keep a backup trace in the password file
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, subkey='foo', overwrite=true, backup=true) }}
" localhost
localhost | SUCCESS => {
    "msg": "8jRMhmn9bE-Ra:Yh"
}
# Replace the main password
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, overwrite=true) }}" localhost
localhost | SUCCESS => {
    "msg": "Mjol:uhPxIiGcchM"
}
# Display the content of the password file
❯ pass example/test 
Mjol:uhPxIiGcchM

foo: 8jRMhmn9bE-Ra:Yh
lookup_pass: First generated by ansible on 01/10/2024 11:37:25
lookup_pass: old subkey password foo was u6V31gj3d0i3lLbO (Updated on 01/10/2024 11:38:03)

```

### CODE REVIEW

File: changelogs/fragments/8952-password-store-lookup-create-subkey-support.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+    - passwordstore lookup - add subkey creation/update support
```
Comment (felixfontein) [inline changelogs/fragments/8952-password-store-lookup-create-subkey-support.yml:2]:
```suggestion
    - passwordstore lookup plugin - add subkey creation/update support (https://github.com/ansible-collections/community.general/pull/8952).
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! A few first comments are below:

The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

File: plugins/lookup/passwordstore.py [hunk 2]
```diff
@@ -525,11 +577,17 @@ def run(self, terms, variables, **kwargs):
             self.parse_params(term)   # parse the input into paramvals
             with self.opt_lock('readwrite'):
                 if self.check_pass():     # password exists
-                    if self.paramvals['overwrite'] and self.paramvals['subkey'] == 'password':
-                        with self.opt_lock('write'):
+                    if self.paramvals['overwrite']:
+                        if self.paramvals['subkey'] == 'password':
+                          with self.opt_lock('write'):
                             result.append(self.update_password())
+                        elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```
Comment (russoz) [inline plugins/lookup/passwordstore.py:584]: Indentation is off.
```suggestion
                            with self.opt_lock('write'):
                                result.append(self.update_password())
                        elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```

File: plugins/lookup/passwordstore.py [hunk 3]
```diff
@@ -525,11 +577,17 @@ def run(self, terms, variables, **kwargs):
...
+                            with self.opt_lock('write'):
+                              result.append(self.update_password())
+                        else:
+                            result.append(self.get_passresult())
                     else:
-                        result.append(self.get_passresult())
+                            result.append(self.get_passresult())
```
Comment (russoz) [inline plugins/lookup/passwordstore.py:590]: The indentation appears to be off here as well, please review.

**Comment by mluzarreta** [comment]
> Thanks for your contribution! A few first comments are below:
> 
> The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

@felixfontein I added some details in the documentation but this PR does not add any parameter, it just allows the creation/update of a subkey using the existing parameters (subkey, create, overwrite, backup).

@felixfontein @russoz I rewrote the commit with your suggestions. Thank you for the review :smile:

**Comment by felixfontein** [comment]
> The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

It seems I totally overlooked that the option already exists. Sorry for that :)

**Comment by mluzarreta** [comment]
FYI I've just corrected a regression:

```
-                    elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
+                    elif self.paramvals["subkey"] != "password" and not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```

**Review by russoz — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this in a week.

**Comment by felixfontein** [comment]
@mluzarreta thanks for your contribution!
@russoz thanks for reviewing!

**Pull request merged**